control-plane/plugin/src/resources/blockdevice.rs

```diff
@@ -0,0 +1,210 @@
+                    .join(", "),
+            ])
+        }
+        row
+    }
+}
+
+fn get_partition_type(partition: &openapi::models::BlockDevicePartition) -> String {
+    if !partition.scheme.is_empty() && !partition.typeid.is_empty() {
+        return format!("{}:{}", partition.scheme, partition.typeid);
+    }
+    "".to_string()
+}
+
+// GetHeaderRow being trait for BlockDeviceAll would return the Header Row for
+// BD returned from REST call with all set to true.
+impl GetHeaderRow for BlockDeviceAll {
+    fn get_header_row(&self) -> Row {
+        (&*BLOCKDEVICE_HEADERS_ALL).clone()
+    }
+}
+
+// GetHeaderRow being trait for BlockDeviceUsable would return the Header Row for
+// BD returned from REST call with all set to false.
+impl GetHeaderRow for BlockDeviceUsable {
+    fn get_header_row(&self) -> Row {
+        (&*BLOCKDEVICE_HEADERS_USABLE).clone()
+    }
+}
+
+#[async_trait(?Send)]
+impl GetBlockDevices for BlockDevice {
+    type ID = NodeId;
+    async fn get_blockdevices(id: &Self::ID, all: &bool, output: &OutputFormat) {
+        if let Err(e) = get_used_devices(id).await {
+            println!("Failed to list blockdevices for node {} . Error {}", id, e)
+        }
+        match RestClient::client()
+            .block_devices_api()
+            .get_node_block_devices(id, Some(*all))
+            .await
+        {
+            Ok(blockdevices) => {
+                // Print table, json or yaml based on output format.
+                if *all {
+                    let bds: Vec<BlockDeviceAll> = blockdevices
+                        .into_body()
+                        .into_iter()
+                        .map(BlockDeviceAll)
+                        .collect();
+                    print_table(output, bds);
+                } else {
+                    let bds: Vec<BlockDeviceUsable> = blockdevices
+                        .into_body()
+                        .into_iter()
+                        .map(BlockDeviceUsable)
+                        .collect();
+                    print_table(output, bds);
+                };
+            }
+            Err(e) => {
+                println!("Failed to list blockdevices for node {} . Error {}", id, e)
+            }
+        }
+    }
+}
+
+async fn get_used_devices(nodeid: &str) -> Result<(), Error<RestJsonError>> {
+    match RestClient::client()
+        .pools_api()
+        .get_node_pools(nodeid)
+        .await
+    {
+        Ok(pools) => {
+            let mut used_disks: Vec<String> = vec![];
+            for pool in pools.into_body() {
+                used_disks.append(&mut pool.spec.unwrap().disks);
+            }
+            USED_DISKS
+                .set(used_disks)
+                .expect("Expected to be initialised once");
+            Ok(())
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn check_available(possible_values: Vec<String>) -> bool {
+    for value in possible_values {
+        if USED_DISKS.get().unwrap().contains(&value) {
```

Comment:
This won't work if the diskpool is created with URI's?

Reply:
Yes, right. Thanks for pointing out. Added normalization.